toc/rfc/rfc-draft-add-cnbs.md

```diff
@@ -0,0 +1,194 @@
+# Meta
+[meta]: #meta
+- Name: Add Cloud Native Buildpacks
+- Start Date: 2023-05-01
+- Author: @gerg
+- Contributing Authors: @ryanmoran, @robdimsdale, @dsabeti, @aramprice, @dsboulder
+- Status: Draft
+- RFC Pull Request: #591
+
+## Summary
+
+[Cloud Native Buildpacks (CNBs)](https://buildpacks.io/), also known as v3
+buildpacks, are the current generation of buildpacks, and offer some
+improvements over the v2 buildpacks that CF Deployment currently uses. The
+Cloud Foundry Foundation already has an implementation of Cloud Native
+Buildpacks via the [Paketo](https://paketo.io/) project, however these CNBs are
+not currently integrated with core CF Deployment.
+
+This RFC proposes adding CNB support to Cloud Foundry and including the Paketo
+buildpacks in CF Deployment.
+
+## Problem
+
+The v2 buildpacks are effectively in maintenance mode and do not receive
+substantial new features. By not integrating with v3 buildpacks, Cloud Foundry
+is missing out on new capabilities like application
+[SBOM](https://en.wikipedia.org/wiki/Software_supply_chain) generation, new
+buildpacks like the Java Native and Web Servers CNBs, as well as any new
+features that are added to the still-actively-developed v3 buildpacks.
+
+## Proposal
+
+### Overview
+
+Cloud Foundry integration with CNBs will be executed over at least two phases.
+
+In the first phase, the Paketo buildpacks will be "shimmed" to work with Cloud
+Foundry's existing v2 buildpack interface. These shimmed Paketo buildpacks will
+be bundled with select [CNB lifecycle](https://github.com/buildpacks/lifecycle)
+binaries and lightweight orchestration binaries. At build time, the Cloud
+Foundry [buildpack lifecycle](https://github.com/cloudfoundry/buildpackapplifecycle)
+will invoke the orchestration binaries, which will map the v2 build process to
+the v3 build process.
+
+In the second phase, unmodified Paketo buildpacks will be natively supported by
+Cloud Foundry. This will be implemented by integrating the buildpack shim logic
+into a Cloud Foundry lifecycle.
```

Comment:
This mean that there are no changes planned for the current V2 [buildpack lifecycle](https://github.com/cloudfoundry/buildpackapplifecycle). I think that we should mention it explicitly to avoid misunderstanding here.

Reply:
I was intentionally non-prescriptive here about whether it would be appropriate to extend the current Buildpack App Lifecycle to support v3 buildpacks, or whether it made sense to create an entirely separate lifecycle (CNB App Lifecycle?).

If desired, we can update the RFC to be more opinionated on that matter.

Reply:
I read it as decided and wanted to make it clear but it is even better that you think for different options here. I agree that it makes sense to decide on this when we have more input from phase 1. Regarding extending the current buildpack lifecycle or having a CNB app lifecycle I think we should go with the approach which makes possible next phases/steps like building OCI images and integration with Container registries easier.  At the moment I'm in favour of a separate CNB app lifecycle.

Reply:
I am not sure if you meant this that way @Gerg, but I agree with @beyhan that we should keep the longer-term in mind and sight when discussing and working on a shorter-term workaround.

I see the benefit of munching Paketo buildpacks and their execution environment (aka CNB lifecycle) together and wrap it up in a v2 compatible shim. Any user can try `cf push -b v3_java_buildpack` on any foundation and give feedback on if and how their app still works.

When starting efforts inside the Cloud Foundry codebase, we should take into account how this can end up in running CNBs natively. We anyway need to run the CNB lifecycle in a Diego scheduled container. We should make sure this stays usable as is and could be pushed into a registry, even if we add logic that extracts a droplet from the result.

For what it's worth, we have written a Cloud Native Buildpack platform with capabilities similar to `pack build` that can run fully inside a container. The [`cnbBuild` step](https://github.com/SAP/jenkins-library/blob/master/cmd/cnbBuild.go) is a bit buried in a larger codebase for a whole CI/CD system / library, but we could certainly help to extract (and maintain) it in a different place or try to push some of the platform logic directly into the `lifecycle`. I've never tried to run in in a Diego Task, but it sure doesn't need a Docker Daemon.

Reply:
TOC discussion:
* follow on RFC required once we've learned lessons (phase 2)

Reply:
https://github.com/cloudfoundry/community/commit/6d6bd816cefc07d71882af16300c93876f8e301a